src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.cs

```diff
@@ -388,7 +388,33 @@ private bool MethodCallOrArrayIndex(AcceptedCharacters acceptedCharacters)
                     }
                     return MethodCallOrArrayIndex(acceptedCharacters);
                 }
-                if (CurrentSymbol.Type == CSharpSymbolType.Dot)
+                if (At(CSharpSymbolType.QuestionMark))
+                {
+                    var next = Lookahead(count: 1);
+
+                    if (next != null)
+                    {
+                        if (next.Type == CSharpSymbolType.Dot)
+                        {
+                            // Accept null propagation dot operator (?.).
+                            AcceptAndMoveNext();
+                            AcceptAndMoveNext();
+
+                            // If the next piece after the ?. is a keyword or identifier then we want to continue.
```

Comment:
How is that not valid? If there's a property which happens to be a delegate named `nameof` it works fine. Same for `class`, `typeof`, `for` etc.